refactor(cache): improve caching system and add expiring cache

nathabonfim59 opened the pull request:
### Description 
For my usage which primarily uses Copilot, this change made a huge difference in performance - reducing about ~2 seconds for each request by caching the access tokens.

Not sure if this is the direction you want to go, but I found it very helpful for my workflow.
Let me know if you'd like any changes - happy to update!

### Changes 

1. Convert existing cache to generic Cache[T] type
   - Split into "conversations" and "temporary" cache types
   - Add better error handling and directory management
   - Improve type safety with generics

2. Add ExpiringCache implementation
   - Support for time-based cache expiration
   - Automatic cleanup of expired entries
   - Used for caching Copilot access tokens

```
$HOME/user/.local/share/mods
.
├── cache
│   └── copilot.1734276998
└── conversations
    ├── 0dbcdaae5bf314e06a9b47bf513e4ab69483761e.gob
    ├── 14a5b9133e384ae4f42c3a49756f6be9703c6982.gob
    ├── 35588af9aeec76b98ee8e4776e6be751e9528417.gob
    └── mods.db
```

**Note:** This PR builds on #406 and should be merged after it.

caarlos0 commented on cache.go:
```diff
@@ -15,58 +18,107 @@ const cacheExt = ".gob"
+
+const (
+	ConversationCache CacheType = "conversations"
+	TemporaryCache    CacheType = "cache"
```
maybe name it "temp" instead of "cache"? or something more specific?

caarlos0 commented:
hey! thanks for the PR! can you merge with main?

thanks!

nathabonfim59 commented:
Done!

Sorry for the `force-pushes` 😅

Had some conflicts with main in my fork and had to rebase my commits to make it work. Also added a wrapper around the errors and fixed some minor linting issues.

If you need me to change anything, just let me know.

caarlos0 commented on cache.go:
```diff
@@ -6,67 +6,124 @@ import (
 }
 
-func (c *convoCache) read(id string, messages *[]openai.ChatCompletionMessage) error {
+func (c *Cache[T]) cacheDir() string {
```
```suggestion
func (c *Cache[T]) dir() string {
```

no nead to repeat cache I think

caarlos0 replied:
done on main

caarlos0 commented:
Thanks @nathabonfim59 for the PRs 🙏🏻
